source/common/config/udpa_resource.cc

```diff
@@ -0,0 +1,79 @@
+#include "common/config/udpa_resource.h"
+
+#include <algorithm>
+
+#include "common/common/fmt.h"
+#include "common/http/utility.h"
+
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_split.h"
+
+// TODO(htuch): This file has a bunch of ad hoc URI encoding/decoding based on Envoy's HTTP util
+// functions. Once https://github.com/envoyproxy/envoy/issues/6588 lands, we can replace with GURL.
+
+namespace Envoy {
+namespace Config {
+
+using PercentEncoding = Http::Utility::PercentEncoding;
+
+std::string UdpaResourceName::encodeUri(const udpa::core::v1::ResourceName& resource_name,
+                                        const EncodeOptions& options) {
+  // We need to percent-encode authority, id, path and query params. Qualified types should not have
+  // reserved characters.
+  const std::string authority = PercentEncoding::encode(resource_name.authority(), "%/#?&=");
```

Comment:
Can authority be empty, to indicate use of some default local authority?  I'm wondering if that would be useful in non-federated use cases.